pulser/tests/test_simresults.py

```diff
@@ -97,45 +103,100 @@ def test_get_final_state():
+    results_single = sim_single.run()
+    op = [qutip.basis(2, 0).proj()]
+    exp = results_single.expect(op)[0]
+    assert np.isclose(exp[-1], 1)
+    assert len(exp) == number_of_meas
+
+
+def test_expect_noisy():
+    bad_op = qutip.tensor([qutip.qeye(2), qutip.sigmap()])
+    with pytest.raises(ValueError, match="non-diagonal"):
+        results_noisy.expect([bad_op])
+    op = qutip.tensor([qutip.qeye(2), qutip.basis(2, 0).proj()])
+    results_noisy.expect([op])
+
+
+def test_plot():
+    op = qutip.tensor([qutip.qeye(2), qutip.basis(2, 0).proj()])
+    results_noisy.plot(op)
+    results_noisy.plot(op, error_bars=False)
+    results.plot(op)
 
 
 def test_sample_final_state():
-    with pytest.raises(ValueError, match="undefined measurement basis"):
-        sim_no_meas = Simulation(seq_no_meas)
-        results_no_meas = sim_no_meas.run()
-        results_no_meas.sample_final_state()
-    with pytest.raises(ValueError, match="can only be"):
-        results_no_meas.sample_final_state('wrong_measurement_basis')
+    sim_no_meas = Simulation(seq_no_meas, config=SimConfig(runs=1))
+    results_no_meas = sim_no_meas.run()
+    results_no_meas.sample_final_state()
     with pytest.raises(NotImplementedError, match="dimension > 3"):
         results_large_dim = deepcopy(results)
         results_large_dim._dim = 7
         results_large_dim.sample_final_state()
 
-    sampling = results.sample_final_state(N_samples=1234)
-    assert results.N_samples == 1234
+    sampling = results.sample_final_state(1234)
     assert len(sampling) == 4  # Check that all states were observed.
 
-    sampling0 = results.sample_final_state(meas_basis='digital', N_samples=911)
+    results._meas_basis = 'digital'
+    sampling0 = results.sample_final_state(N_samples=911)
     assert sampling0 == {'00': 911}
-
     seq_no_meas.declare_channel('raman', 'raman_local', 'B')
     seq_no_meas.add(pi, 'raman')
     res_3level = Simulation(seq_no_meas).run()
-    sampling_three_level = res_3level.sample_final_state(meas_basis='digital')
     # Raman pi pulse on one atom will not affect other,
     # even with global pi on rydberg
-    assert len(sampling_three_level) == 2
-    sampling_three_levelB = res_3level.sample_final_state(
-                                meas_basis='ground-rydberg')
-    # Global Rydberg will affect both:
+    assert len(res_3level.sample_final_state()) == 2
+    res_3level._meas_basis = 'ground-rydberg'
+    sampling_three_levelB = res_3level.sample_final_state()
+    # Rydberg will affect both:
     assert len(sampling_three_levelB) == 4
+
+
+def test_sample_final_state_noisy():
+    results_noisy.sample_final_state(N_samples=1234)
+    results_noisy.sample_final_state(N_samples=911)
+    seq_no_meas_noisy.declare_channel('raman', 'raman_local', 'B')
+    seq_no_meas_noisy.add(pi, 'raman')
+    res_3level = Simulation(seq_no_meas_noisy, config=SimConfig(
+        noise=('SPAM', 'doppler'), runs=1))
+    res_3level.run().states
```

Comment:
Assert the outcome is what you expect (you can use `results` instead of `states` if it's more convenient)